### ISSUE TITLE
V2 Reusing tests: generating tests with factories

### ISSUE THREAD

**Issue by Jelleas** [Jelleas/CheckPy#28]
**TL:DR;**

```Py
testAppleDef = helpers.assertDefFactory(function="simulate_apple", parameters=["x", "dt"])
```

**Problem:** checks with descriptive error messages, hints and detailed asserts are both lengthy and hard to re-use. For instance:

```Py
from checkpy import *

@test()
def testAppleDef():
	"""Defines the function `simulate_apple()`"""
	assert "simulate_apple" in static.getFunctionDefinitions(),\
		"make sure the function is defined"

	f = getFunction("simulate_apple")

	assert f.parameters == ["x", "dt"],\
		"make sure the function has the correct parameters"
```

Re-using this check would require choosing a new test name, a new description, carefully replacing the function name "simulate_apple()" in multiple places.

**Solution:** a factory-style approach to generating tests. In the same file or in a separate file (for instance `helpers.py`):

```Py
import checkpy
from uuid import uuid4

def assertDefFactory(
    function: str,
    parameters: Tuple[Any]=(),
    timeout: int=10,
) -> checkpy.tests.Test:
    """Generates a test to assert a function definition is present with set parameters."""
    def foo():
        assert function in checkpy.static.getFunctionDefinitions(),\ 
            "make sure the function is defined with the correct name"

        f = checkpy.getFunction(function)

        assert f.parameters == list(parameters),\
            "make sure the function has the correct parameters"
 
    # Give the test function a unique name
    # Normally this would be the test's function name, but we don't have that here
    foo.__name__ == str(uuid4())

    # Define the description separately here
    # A formatted docstring  (f"""""") is only evaluated when the function is called
    # And... that might not happen if the test depends on others.
    foo.__doc__ = f"Defines the function `{function}()`"

    # Call the test decorator (checkpy.test()), then pass the test function foo
    return checkpy.test(timeout=timeout)(foo)
```

Back in the tests file replace the original test (`testAppleDef`) with:

```Py
import helpers

testAppleDef = helpers.assertDefFactory(function="simulate_apple", parameters=["x", "dt"])
```

In case `helpers.py` is defined in another directory, say the parent directory of the test. Be sure to add that directory to `sys.path`. For instance:

```Py
import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).parent))
import helpers
```

### What about dependencies?

```Py
# Call `passed` or `failed` directly
testAppleDef = passed(someOtherTest)(helpers.assertDefFactory(function="simulate_apple", parameters=["x", "dt"]))

# Other checks can depend on generated checks like normal
@passed(testAppleDef)
def testAppleFunc():
     ...
```

### What about custom hints and checks?

Probably best to add the option for a callback to the factory. For instance:

```Py
def assertDefFactory(
    ...
    before: Callable[[], None]=lambda: None,
) -> checkpy.tests.Test:
    """Generates a test to assert a function definition is present with set parameters."""
    def foo():
        before()
        ...
```

Then use it like so:

```Py
testAppleDef = helpers.assertDefFactory(
	function="simulate_apple",
	parameters=["x", "dt"],
	before=lambda: helpers.checkForNotAllowedCode("break")
)
```

### Loops?

Sure.

```Py
for name, params in [("foo", ["x"]), ("bar", ["x", "y"])]:
     globals()[f"testDef{name.capitalize()}"] =\
        helpers.assertDefFactory(function=name, parameters=params)
```

### More examples?

```Py
def assertFuncFactory(
    function: str,
    input: Union[Tuple[Any], Dict[str, Any]],
    output: Any,
    outputType: type=Any,
    hint: Callable[[Any], None]=lambda output: None,
    timeout: int=10,
) -> checkpy.tests.Test:
    """Generates a test to assert the outcome of a function call."""
    def foo():
        f = checkpy.getFunction(function)

        if isinstance(input, dict):
            inputStr = ", ".join([f"{k}={v}" for k, v in input.items()])
        else:
            inputStr = ", ".join(str(v) for v in input)

        if isinstance(input, dict):
            realOutput = f(**input)
        else:
            realOutput = f(*input)

        assert realOutput == checkpy.Type(outputType),\
            "make sure the function returns the correct type"

        hint(realOutput)

        assert realOutput == output,\
            f"Did not expect output {realOutput} (with input {inputStr})"

    foo.__name__ == str(uuid4())
    parameterStr = ", ".join(str(p) for p in input)
    foo.__doc__ = f"Testing {function}({parameterStr})"
    return checkpy.test(timeout=timeout)(foo)
```

```Py
def appleHint(output):
	t, v = output
	if v == approx(4.52, abs=0.1) or t == approx(159.47, abs=1):
		assert False, "Did you mix up the order of the return values?"

	if v == approx(44.3, abs=0.3):
		assert False, "Did you forget to convert to km/h?"

testApple = passed(testAppleDef, timeout=90, hide=False)(
	helpers.assertFuncFactory(
		function="simulate_apple",
		input=(100, 0.01),
		output=(approx(159.47, abs=0.1), approx(4.52, abs=0.1)),
		outputType=Tuple[float, float],
		hint=appleHint
	)
)
```

**Comment by Jelleas** [comment]
Alternative approach using method chaining:

```Py
def appleHint(output):
    t, v = output
    if v == approx(4.52, abs=0.1) or t == approx(159.47, abs=1):
        assert False, "Did you mix up the order of the return values?"

    if v == approx(44.3, abs=0.3):
        assert False, "Did you forget to convert to km/h?"

testApple = (helpers.FuncTestBuilder("simulate_apple")
    .params("x", "dt")
    .returnType(Tuple[float, float])
    .call(100, 0.01),
    .do(applehint)
    .returned((approx(159.47, abs=0.1), approx(4.52, abs=0.1)))
    .resolve()
)
```
